File: LeetCode/mergeSort.cpp
//Divide and Conquer Algorithm
//Best and AVeage time O(nlogn)
//Worst case time O(nlogn)
//Requires extra memory for temporary arrays

void merge(int arr[], int l, int m, int r) {
	int i, j, k;
	int n1 = m - l + 1;
	int n2 = r - m;
	int leftTemp[n1];
	int rightTemp[n2];

	for (i = 0; l < m; l++) {
		leftTemp[i] = arr[l+i];
	}
	for (j = 0; j < r; j++) {
		leftTemp[j] = arr[l+m+j];
	}

	while (i < n1 && j < n2) {
		if (leftTemp[i] <= rightTemp[j]) {
			arr[k] = leftTemp[i];
			i++;
		}
		else {
			arr[k]= rightTemp[j];
			j++;
		}
		k++;
	}

	while (i < n1) {
		arr[k] = arr[i];
		i++;
		k++;
	}

	while (j < n2) {
		arr[k] = arr[j];
		j++;
		k++;
	}
}

void mergeSort(int arr[], int l, int r) {
	if (l > r) {
		int middle = (l+r)/2;
		mergeSort(int arr[], l, middle);
		mergeSort(int arr[], middle+ + 1, r)

		merge(int arr[], l, middle, r);
	}
}

File: LeetCode/mergeSort.py

mergesort(array):
	if len(array) <= 1:
		return array

	middle = len(array)//2
	left = mergesort(array[:middle])
	right = mergesort(arra[middle:])
	merge(left, right)

merge(left, right):
	leftLen = len(left)
	rightLen = len(right)
	result = []
	i = 0
	j = 0
	while i < leftLen and j < rightLen:
		if left[i] <= right[j]:
			result.append(left[i])
			i+=1
		else:
			result.append(right[j])
			j+=1

	result += left[i:]
	result += right[j:]
	return result


